File: SmartHomeWeb.Server/Services/RoomService.cs
public class RoomService
{
    private readonly HomeService _homeService;
    private readonly UserService _userService;

    public RoomService(UserService userService, HomeService homeService)
    {
        _userService = userService;
        _homeService = homeService;
    }

    public async Task AddRoomAsync(string userId, string homeId, Room room)
    {
        var user = await _userService.GetUserByIdAsync(userId);
        var home = user.Homes.Find(h => h.HomeId == homeId);
        home.Rooms.Add(room);
        await _userService.UpdateUserAsync(userId, user);
    }

    public async Task<List<Room>> GetRoomsByHomeIdAsync(string userId, string homeId) =>
        (await _userService.GetUserByIdAsync(userId))?.Homes.Find(h => h.HomeId == homeId)?.Rooms ?? new List<Room>();

    public async Task DeleteRoomAsync(string userId, string homeId, string roomId)
    {
        var user = await _userService.GetUserByIdAsync(userId);
        var home = user.Homes.Find(h => h.HomeId == homeId);
        home.Rooms.RemoveAll(r => r.RoomId == roomId);
        await _userService.UpdateUserAsync(userId, user);
    }
}

File: smarthomeweb.client/src/app/services/room.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Device } from './device.service'
import { Observable } from 'rxjs';

export interface Room {
  roomId: string;
  name: string;
  devices: Device[];
}

@Injectable({
  providedIn: 'root',
})
export class RoomService {
  private apiUrl = 'https://web-app-t5-dev-aca2dahff0bkb5g9.westeurope-01.azurewebsites.net/api/User';

  constructor(private http: HttpClient) { }

  addRoom(userId: string, homeId: string, room: Room): Observable<any> {
    return this.http.post(`${this.apiUrl}/${userId}/homes/${homeId}/rooms`, room);
  }

  getRoomsByHomeId(userId: string, homeId: string): Observable<Room[]> {
    return this.http.get<Room[]>(`${this.apiUrl}/${userId}/homes/${homeId}/rooms`);
  }
}
